core/utils/src/main/kotlin/edu/stanford/spezi/core/utils/foundation/SharedRepository.kt

```diff
@@ -0,0 +1,97 @@
+package edu.stanford.spezi.core.utils.foundation
+
+import edu.stanford.spezi.core.utils.foundation.knowledgesource.ComputedKnowledgeSource
+import edu.stanford.spezi.core.utils.foundation.knowledgesource.ComputedKnowledgeSourceStoragePolicy
+import edu.stanford.spezi.core.utils.foundation.knowledgesource.DefaultProvidingKnowledgeSource
+import edu.stanford.spezi.core.utils.foundation.knowledgesource.KnowledgeSource
+import edu.stanford.spezi.core.utils.foundation.knowledgesource.OptionalComputedKnowledgeSource
+import kotlin.reflect.KClass
+
+@Suppress("detekt:TooManyFunctions")
+interface SharedRepository<Anchor : RepositoryAnchor> {
```

Comment:
Bear with me as a long comment ahead with lots of code 😄 . I will add my full proposal here. As proposed above, I see `RepositoryAnchor` as redundant, while we can achieve the same with a new marker interface which can indeed be useful for all calculations. And an additional gain is working with one single generic type instead of two as it is currently the case. Results in a simplier API surface imo. As this implies deviation from iOS by removing the `RepositoryAnchor` interface, I would propose to first align with @PSchmiedmayer and @Supereg beforehand. We will start with a new marker interface:

```kotlin
// Serves as a link (similar to our existing anchor) to the Repository
interface RepositoryValue
``` 

Next, we need to achieve the link between the knowledge sources, and the values that they address, hence our knowledge sources which serve as a key to access store and retrieve values of type `RepositoryValue`. This can be achieved via:

```kotlin
// Knowledge sources
interface KnowledgeSource<Value : RepositoryValue>
```

Let's define the storage policies as you are already doing:

```kotlin
// Knowledge source storage policy
sealed interface ComputedKnowledgeSourceStoragePolicy {
    data object AlwaysCompute : ComputedKnowledgeSourceStoragePolicy
    data object Store : ComputedKnowledgeSourceStoragePolicy
}
```

Including `Shared` in the name feels like an indication / enforcement for the clients on how to use the repositories, hence I would simply go with `Respository` for the name. We need to achieve the link between the repository and the type of values that it can contain, we can do that as follows:

```kotlin
interface Repository<Value : RepositoryValue> {
    // api methods will follow below
}
```

Next, we can define our more refined knowledge sources (note the generic constraint `Value: RepositoryValue` and how the corresponding repository is typed correctly in the `compute` methods).

```kotlin
interface DefaultProvidingKnowledgeSource<Value : RepositoryValue> : KnowledgeSource<Value> {
    val defaultValue: Value
}

interface SomeComputedKnowledgeSource<Value : RepositoryValue> : KnowledgeSource<Value> {
    val storagePolicy: ComputedKnowledgeSourceStoragePolicy
}

interface ComputedKnowledgeSource<Value : RepositoryValue> : SomeComputedKnowledgeSource<Value> {
    fun <V : Value> compute(repository: Repository<in V>): V
}

interface OptionalComputedKnowledgeSource<Value : RepositoryValue> : SomeComputedKnowledgeSource<Value> {
    fun <V : Value> compute(repository: Repository<in V>): V?
}
```

Next, we can we can define our repository api including the default methods:

```kotlin
/**
 * Repository - Generic constraint Repository value - A repository that can contain values of type Value
 */
interface Repository<Value : RepositoryValue> {

    // Start: Base requirements for conforming types
    operator fun <V : Value> get(source: KnowledgeSource<V>): V?
    operator fun <V : Value> set(source: KnowledgeSource<V>, value: V)
    fun <V : Value> collect(allOf: KClass<V>): List<V>
    // End: Base requirements for conforming types

    // Start: Default implementations using base requirements
    fun <V : Value> getOrDefault(
        source: KnowledgeSource<V>,
        default: V,
    ): V {
        return this[source] ?: default
    }

    fun <V : Value> getComputed(
        source: ComputedKnowledgeSource<V>,
    ): V {
        return get(source) ?: when (source.storagePolicy) {
            is ComputedKnowledgeSourceStoragePolicy.AlwaysCompute -> {
                source.compute(this)
            }

            is ComputedKnowledgeSourceStoragePolicy.Store -> {
                val value = source.compute(this)
                this[source] = value
                value
            }
        }
    }

    fun <V : Value> getOrOptionalComputed(
        source: OptionalComputedKnowledgeSource<V>,
    ): V? {
        return this[source] ?: when (source.storagePolicy) {
            is ComputedKnowledgeSourceStoragePolicy.AlwaysCompute -> {
                source.compute(this)
            }

            is ComputedKnowledgeSourceStoragePolicy.Store -> {
                val value = source.compute(this)
                value?.let { this[source] = it }
                value
            }
        }
    }

    fun <V : Value> getOrProvideDefault(defaultSource: DefaultProvidingKnowledgeSource<V>): V {
        return getOrDefault(defaultSource, defaultSource.defaultValue)
    }

    fun <V : Value> contains(source: KnowledgeSource<V>): Boolean {
        return this[source] != null
    }
    // End: Default implementations using base requirements
}
```

What `V : Value` achieves in the api methods, is enabling clients to work with base values and more refined ones, e.g. sealed interfaces (see my example below).

We can go a step further and provide a default implementation similarly as you are doing with `ValueRepository`:

```kotlin
open class ValueRepository<Value : RepositoryValue>(
    // built in thread-safeness for read and write
    private val storage: ConcurrentHashMap<KnowledgeSource<*>, Value> = ConcurrentHashMap(),
) : Repository<Value> {
    final override fun <V : Value> get(source: KnowledgeSource<V>): V? {
        @Suppress("UNCHECKED_CAST")
        return storage[source] as? V
    }

    final override fun <V : Value> set(source: KnowledgeSource<V>, value: V) {
        storage[source] = value
    }

    final override fun <V : Value> collect(allOf: KClass<V>): List<V> {
        return storage.values.filterIsInstance(allOf.java)
    }
}
```

Next, a full example. Let's assume we want to build a repository where we can store our app settings, e.g. theme of the app, language settings and notifications. All of these values are candidates for `RepositoryValue`. We can declare as follows:

```kotlin
sealed interface AppSettings : RepositoryValue {
    sealed interface Theme : AppSettings {
        data object DarkMode : Theme
        data object LightMode : Theme
    }

    sealed interface Language : AppSettings {
        data object German : Language
        data object English : Language
        data class Other(val value: String) : Language
    }
    
    data class Notifications(val enabled: Boolean) : AppSettings
}
```

Next, we define the keys (knowledge sources) to interact with these values (I am keeping it simple and going simply with the base `KnowledgeSource` interface.

```kotlin
sealed interface AppSettingSources<V : AppSettings> : KnowledgeSource<V> {
    data object Language : AppSettingSources<AppSettings.Language>
    data object Theme : AppSettingSources<AppSettings.Theme>
    data object Notifications : AppSettingSources<AppSettings.Theme>
}
```

Next, making use of our value repository, we can declare repository class that will be used for managing app settings as follows:

```kotlin
class AppSettingsRepository : ValueRepository<AppSettings>()
```

and here an example usage in the tests:

```kotlin
@Test
fun testApiUsage() {
    val repository = AppSettingsRepository()
    
    // storing theme
    repository[AppSettingSources.Theme] = AppSettings.Theme.DarkMode
    
    // reading language
    val languageOrDefault = repository.getOrDefault(
        source = AppSettingSources.Language,
        default = AppSettings.Language.Other("Italian")
    )
    
    // storing notifications
    repository[AppSettingSources.Notifications] = AppSettings.Notifications(enabled = false)
    
    // asserting storage and default
    assertThat(languageOrDefault).isEqualTo(AppSettings.Language.Other("Italian"))
    assertThat(repository[AppSettingSources.Theme]).isEqualTo(AppSettings.Theme.DarkMode)
    
    // storing new language
    repository[AppSettingSources.Language] = AppSettings.Language.German
    
    // Asserting via collect
    assertThat(repository.collect(allOf = AppSettings.Theme::class)).isEqualTo(listOf(AppSettings.Theme.DarkMode))
    assertThat(repository.collect(allOf = AppSettings.Language::class)).isEqualTo(listOf(AppSettings.Language.German))
    assertThat(repository.collect(allOf = AppSettings::class)).isEqualTo(
        listOf(
            AppSettings.Theme.DarkMode,
            AppSettings.Language.German,
            AppSettings.Notifications(enabled = false)
        )
    )
}
```

Reply:
The sealed interface in this case would exactly seal the possible values, which is exactly the purpose of this construct to even exist - so that users can add their own keys/values. In addition to that, a value type may be used for multiple keys in the same application area (i.e. with the same anchor) and also a value type is not necessarily limited to one application area (not talking about keys here, just the value type). So, this would actually limit us here.

Reply:
as discussed, since Kotlin does not support interface conformances with the use of extensions, which would mean that we would need to wrap every value that we store in a repository in a new type conforming to that interface, we keep the anchor generic constraint for now - also to stay more aligned with iOS